pkg/license/license.go

```diff
@@ -0,0 +1,311 @@
+// Google's cicense classifier
+type ReaderDefaultImpl struct {
+	lc   *licenseclassifier.Classifier
+	spdx *SPDX
+}
+
+// Initialize checks the options and creates the needed objects
+func (d *ReaderDefaultImpl) Initialize(opts *ReaderOptions) error {
+	// Validate our options before startin
+	if err := opts.Validate(); err != nil {
+		return errors.Wrap(err, "validating the license reader options")
+	}
+
+	// Create the implementation's SPDX object
+	spdxopts := DefaultSPDXOpts
+	spdxopts.CacheDir = opts.CachePath()
+	spdx, err := NewSPDXWithOptions(spdxopts)
+	if err != nil {
+		return errors.Wrap(err, "creating SPDX object")
+	}
+	d.spdx = spdx
+
+	if err := d.spdx.LoadLicenses(); err != nil {
+		return errors.Wrap(err, "loading licenses")
+	}
+
+	// Write the licenses to disk as th classifier will need them
+	if err := spdx.WriteLicensesAsText(opts.LicensesPath()); err != nil {
+		return errors.Wrap(err, "writing license data to disk")
+	}
+
+	// Create the implementation's classifier
+	d.lc = licenseclassifier.NewClassifier(opts.ConfidenceThreshold)
+	return errors.Wrap(d.lc.LoadLicenses(opts.LicensesPath()), "loading licenses at init")
+}
+
+// Classifier returns the license classifier
+func (d *ReaderDefaultImpl) Classifier() *licenseclassifier.Classifier {
+	return d.lc
+}
+
+// SPDX returns the reader's SPDX object
+func (d *ReaderDefaultImpl) SPDX() *SPDX {
+	return d.spdx
+}
+
+// ClassifyFile takes a file path and returns the most probable license tag
+func (d *ReaderDefaultImpl) ClassifyFile(path string) (licenseTag string, moreTags []string, err error) {
+	file, err := os.Open(path)
+	if err != nil {
+		return licenseTag, nil, errors.Wrap(err, "opening file for analysis")
+	}
+	defer file.Close()
+
+	// Get the classsification
+	matches, err := d.Classifier().MatchFrom(file)
+	if len(matches) == 0 {
+		logrus.Warn("File does not match a known license: " + path)
+	}
+	var highestConf float64
+	moreTags = []string{}
+	for _, match := range matches {
+		if match.Confidence > highestConf {
+			highestConf = match.Confidence
+			licenseTag = match.Name
+			moreTags = append(moreTags, match.Name)
+		}
+	}
+	return licenseTag, []string{}, nil
+}
+
+// ClassifyLicenseFiles takes a list of paths and tries to find return all licenses found in it
+func (d *ReaderDefaultImpl) ClassifyLicenseFiles(paths []string) (
+	licenseList []ClassifyResult, unrecognizedPaths []string, err error) {
+	// Run the files through the clasifier
+	for _, f := range paths {
+		label, _, err := d.ClassifyFile(f)
+		if err != nil {
+			return nil, unrecognizedPaths, errors.Wrap(err, "classifying file")
+		}
+		if label == "" {
+			unrecognizedPaths = append(unrecognizedPaths, f)
+			continue
+		}
+		// Get the license corresponding to the ID label
+		license := d.spdx.GetLicense(label)
+		if license == nil {
+			return nil, unrecognizedPaths, errors.Wrapf(nil, "ID does not correspond to a valid license: %s", label)
```

Comment:
Wrapping a nil error will always result in nil.

Reply:
Ooof nice catch. Thanks!